writer/writer.go

```diff
@@ -15,27 +15,380 @@
+	"type-info":       {Path: "type-info", IsDir: false, Required: true},
+	"checksums":       {Path: "checksums", IsDir: true, Required: false},
+	"checksums/*":     {Path: "checksums", IsDir: false, Required: false},
+	"signatures":      {Path: "signatures", IsDir: true, Required: true},
+	"signatures/*":    {Path: "signatures", IsDir: false, Required: true},
+	"scripts":         {Path: "scripts", IsDir: true, Required: false},
+	"scripts/pre":     {Path: "scripts/pre", IsDir: true, Required: false},
+	"scripts/pre/*":   {Path: "scripts/pre", IsDir: false, Required: false},
+	"scripts/post":    {Path: "scripts/post", IsDir: true, Required: false},
+	"scripts/post/*":  {Path: "scripts/post", IsDir: false, Required: false},
+	"scripts/check":   {Path: "scripts/check", IsDir: true, Required: false},
+	"scripts/check/*": {Path: "scripts/check/*", IsDir: false, Required: false},
+	// we must have data directory containing update
+	"data":   {Path: "data", IsDir: true, Required: true},
+	"data/*": {Path: "data/*", IsDir: false, Required: true},
+}
+
+// NewArtifactsWriter creates a new ArtifactsWriter providing a location
+// of Mender metadata artifacts, format of the update and version.
+func NewArtifactsWriter(path string, format string, version int) *ArtifactsWriter {
+	return &ArtifactsWriter{
+		updateLocation:  path,
+		headerStructure: metadata.MetadataArtifactHeader{Artifacts: ArtifactsHeaderFormat},
+		format:          format,
+		version:         version,
+		updates:         make(map[string]updateBucket),
+	}
+}
+
+type updateArtifact struct {
+	name         string
+	path         string
+	updateBucket string
+	info         os.FileInfo
+	checksum     []byte
+}
+
+type updateBucket struct {
+	location        string
+	path            string
+	archivedPath    string
+	updateArtifacts []updateArtifact
+	files           metadata.MetadataFiles
+}
+
+// ReadArchiver provides interface for reading files or streams and preparing
+// those to be written to tar archive.
+// GetHeader returns Header to be written to the crrent entry it the tar archive.
+type ReadArchiver interface {
+	io.ReadCloser
+	Open() error
+	GetHeader() (*tar.Header, error)
+}
+
+func (av ArtifactsWriter) calculateChecksum(upd *updateArtifact) error {
+	f, err := os.Open(upd.path)
+	if err != nil {
+		return errors.Wrapf(err, "can not open file for calculating checksum")
+	}
+	defer f.Close()
+	h := sha256.New()
+	if _, err := io.Copy(h, f); err != nil {
+		return errors.Wrapf(err, "error calculating checksum")
+	}
+
+	checksum := h.Sum(nil)
+	upd.checksum = make([]byte, hex.EncodedLen(len(checksum)))
+	hex.Encode(upd.checksum, h.Sum(nil))
+	log.Debugf("hash of file: %v (%x)\n", upd.path, upd.checksum)
+	return nil
+}
+
+func (av ArtifactsWriter) getHeaderInfo(updates []os.FileInfo) metadata.MetadataHeaderInfo {
+	// for now we have ONLY one type of update - rootfs-image
+	headerInfo := metadata.MetadataHeaderInfo{}
+
+	// TODO: should we store update name as well?
```

Comment:
I have been more thinking that instead of: 
```
{
  "updates": [
    {
      "type": "rootfs-image"
    },
    {
      "type": "rootfs-image"
    },
    {
      ...
    }
  ]
}
```
we can have:
```
{
  "updates": [
    {
      "type": "rootfs-image"
      "update": "0000/core-image-minimal-201608110900.ext4"
    },
    {
      "type": "rootfs-image"
      "update": "0000/core-image-base-201609010900.ext3"
    },
    {
      ...
    }
  ]
}
```
Having that maybe we can get rid of some other files (like files), or combine files and type-info.